onnxruntime/core/graph/graph.cc

```diff
@@ -3426,6 +3426,33 @@ void Graph::CleanUnusedInitializers(const std::unordered_set<std::string>* initi
                     }
                   }
                 });
+
+  // Clear the unused NodeArgs
+  // We also want to scan the output NodeArgs of each node
+  // In case one output of a node is neither used as an input of another node nor an output of graph
+  for (const auto& node : Nodes()) {
+    for (const auto* def : node.OutputDefs()) {
+      ORT_IGNORE_RETURN_VALUE(used_args.insert(def->Name()));
+    }
+  }
+
+  // We also need to check the Outer Scope NodeArgs
+  for (const auto& outer_scope_node_arg_name : outer_scope_node_arg_names_) {
+    ORT_IGNORE_RETURN_VALUE(used_args.insert(outer_scope_node_arg_name));
+  }
+
+  for (auto it = node_args_.cbegin(), node_args_end = node_args_.cend(); it != node_args_end; /* no increment */) {
+    auto current_entry = it++;
+    const auto& node_arg_name = current_entry->first;
+    if (!node_arg_name.empty() && used_args.find(node_arg_name) == end &&
+        (initializer_names_to_preserve == nullptr ||
```

Comment:
if we are matching the NodeArg names too, the name `initializer_names_to_preserve` seems less precise now

Reply:
Moved to a new std::unordered_set `node_args_to_preserve` to use NodeArg* as the key